utils/s3/client.cc

```diff
@@ -854,6 +861,209 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
+        auto req = http::request::make("DELETE", _client->_host, _object_name);
+        req.query_parameters["uploadId"] = std::exchange(_upload_id, "");
+        co_await _client->make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
+    }
+
+    class throttled_consumer {
+        output_stream<char>& _output;
+        client& _client;
+        using consumption_result_type = consumption_result<char>;
+
+    public:
+        throttled_consumer(output_stream<char>& output, client& client)
+            : _output{output}
+            , _client{client} {
+        }
+        future<consumption_result_type> operator()(temporary_buffer<char> buf) {
+            if (buf.empty()) {
+                co_return stop_consuming(std::move(buf));
+            }
+            auto mem_units = co_await _client.claim_memory(buf.size());
+            co_await _output.write(buf.get(), buf.size());
+            co_return continue_consuming();
+        }
+    };
+
+    future<> upload_part(file&& f, uint64_t offset, uint64_t part_size) {
+        // upload a part in a multipart upload, see
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        unsigned part_number = _part_etags.size();
+        _part_etags.emplace_back();
+        auto req = http::request::make("PUT", _client->_host, _object_name);
+        req._headers["Content-Length"] = to_sstring(part_size);
+        req.query_parameters.emplace("partNumber", to_sstring(part_number + 1));
+        req.query_parameters.emplace("uploadId", _upload_id);
+        s3l.trace("PUT part {}, {} bytes (upload id {})", part_number, part_size, _upload_id);
+        req.write_body("bin", part_size, [f=std::move(f), offset, part_size, this] (output_stream<char>&& out_) mutable -> future<> {
+            auto output = std::move(out_);
+            auto input = make_file_input_stream(std::move(f), offset, part_size);
+            co_await input.consume(throttled_consumer{output, *_client});
+            co_await output.close();
+            co_await input.close();
+        });
+        return _client->make_request(std::move(req), [this, part_size, part_number, start = s3_clock::now()] (group_client& gc, const http::reply& reply, input_stream<char>&& in_) mutable -> future<> {
+            auto etag = reply.get_header("ETag");
+            s3l.trace("uploaded {} part data -> etag = {} (upload id {})", part_number, etag, _upload_id);
+            _part_etags[part_number] = std::move(etag);
+            gc.write_stats.update(part_size, s3_clock::now() - start);
+            return make_ready_future();
+        }).handle_exception([this, part_number] (auto ex) {
+            s3l.warn("couldn't upload part {}: {} (upload id {})", part_number, ex, _upload_id);
+        });
+    }
+
+    static size_t calc_part_size(size_t total_size, size_t part_size) {
+        if (part_size > 0) {
+            assert(part_size >= aws_minimum_part_size);
```

Comment:
The part_size is some input parameter from the caller, if it's "wrong" the method should throw. If we think that user mustn't pass "wrong" part_sizes, then `on_internal_error()` is still better than assert